llvm/lib/Target/NVPTX/NVPTXAssignValidGlobalNames.cpp

```diff
@@ -73,10 +74,10 @@ std::string NVPTXAssignValidGlobalNames::cleanUpName(StringRef Name) {
   std::string ValidName;
   raw_string_ostream ValidNameStream(ValidName);
   for (char C : Name) {
-    if (C == '.' || C == '@' || C == '<' || C == '>') {
-      ValidNameStream << "_$_";
-    } else {
+    if (isAlnum(C) || C == '_' || C == '$') {
       ValidNameStream << C;
+    } else {
+      ValidNameStream << "_$_";
```

Comment:
Ugh. I think we'll have a problem when we have symbols that have different invalid characters in the same place. E.g. `str-` and `str.`. 
I think ideally we need to encode specific symbols instead. So, instead of just `_$_` for all invalid characters, emit ` "$"+hex(char)` which will be unambiguous.

Reply:
`setName` will add a postfix number to avoid collisions, so I think correctness is fine. Do you think adding the hex values in necessary for readability? I agree the current renaming leaves much to be desired but the hex scheme suggested will also have ambiguous cases. (i.e. `str.` and `str$2E`)

Reply:
> setName will add a postfix number to avoid collisions

That would make such escaping inconsistent between TUs which may be a problem for public symbols that need to match across TUs.

Reply:
This renaming is already limited to names with local linkage.

Reply:
OK. I think we do have some bugs open for generating invalid global PTX identifiers, but that's a bigger, and somewhat different can of worms.

Reply:
Sounds good, I agree there are a lot of unaddressed edge cases here but that's beyond the scope of this PR